### PR TITLE
[Experimental] if_succeeds annotations on statements

### PR DESCRIPTION
**Pull request by cd1m0**
**[WARNING]: This feature is experimental, and may be changed and potentially removed in the future. Feel free to explore it, but please don't depend on it for core annotations until its stabilized.**

This is a Proof-of-concept to show extending `if_succeeds` annotations to statements and blocks inside a function. In these statement level annotations you can talk about the values of local variables before and after a statement:

```
/// #if_succeeds i == old(i)+1;
i = i + 1;
```

and value of variables before and after the body of a loop, or any other code block:

```
for (uint i = 0; i < a.length; i++)
/// #if_succeeds a[i] == old(a[i]) + 1;
{
    a[i] = a[i] + 1; 
}
```

This PR does the following code changes and cleanup:
- Modify `validateAnnotation` to allow `if_succeeds` on statement
- Simplify the `InstrumentationSiteType` enum to just 3 types
- Modify `instrumentStatement` to support 2 types of statement intrumentation - "single point" instrumentation (`assert`, `require`) and "two point" instrumentation (`if_succeeds`). `if_succeeds` is a "two point" instrumentation because it may insert code both before and after the statement.
- Add support for the following edge-case in typechecking:

```
/// #if_succeeds i == 1;
uint i = 1;
```

- Specifically when we have `if_succeeds` on a variable declaration, we need to type-check expressions outside of `old()` and those in `old()` differently, since before the statement there are fewer local varaibles defined. To do this the `STypingCtx` context thats passed around in typing is changed into an object that keeps track of the annotation target, wether the expression is inside of an `old()` and the annotation type.

### CODE REVIEW

File: src/instrumenter/instrument.ts [hunk 1]
```diff
@@ -1139,36 +1139,54 @@ export function instrumentStatement(
...
+    const transCtx = ctx.transCtxMap.get(fun, InstrumentationSiteType.Custom);
 
-    transCtx.resetMarkser([assertionBlock, "end"], false);
+    transCtx.resetMarker([beforeStmtBlock, "end"], false);
 
-    insertAnnotations(allAnnotations as PropertyMetaData[], transCtx);
+    insertAnnotations(singlePointAnnots as PropertyMetaData[], transCtx);
+
+    if (ifSucceedsAnnots.length > 0) {
+        const afterStmtBlock = gte(ctx.compilerVersion, "0.8.0")
```
Comment (cd1m0) [inline src/instrumenter/instrument.ts:1181]: Code duplication with line 1163 - could be a helper on ScribbleFactory?

File: src/instrumenter/transpiling_context.ts [hunk 2]
```diff
@@ -194,35 +192,40 @@ export class TranspilingContext {
...
-            if (
-                this.instrSiteType === InstrumentationSiteType.FunctionAnnotation ||
-                this.instrSiteType === InstrumentationSiteType.StateVarUpdated
-            ) {
-                const uncheckedBlock = this.factory.makeUncheckedBlock([]);
-                this.uncheckedBlocks.push(uncheckedBlock);
-                containerBody.insertAtBeginning(uncheckedBlock);
-
-                this.oldMarkerStack = [[uncheckedBlock, "end"]];
+        /// TODO(dimo): This logic is convoluted and doesn't belong in the constructor. Instead perhaps we should have
```
Comment (cd1m0) [inline src/instrumenter/transpiling_context.ts:195]: Address TODO

File: src/instrumenter/transpiling_context.ts [hunk 3]
```diff
@@ -194,35 +192,40 @@ export class TranspilingContext {
...
-            if (
-                this.instrSiteType === InstrumentationSiteType.FunctionAnnotation ||
-                this.instrSiteType === InstrumentationSiteType.StateVarUpdated
-            ) {
-                const firstStmt = containerBody.vStatements[0];
-
-                this.oldMarkerStack = [[containerBody, ["before", firstStmt]]];
-            }
-
-            this.newMarkerStack = [[containerBody, "end"]];
```
Comment (cd1m0) [inline src/instrumenter/transpiling_context.ts:225]: Should assert that this.instrSiteType === InstrumentationSiteType.Custom to guard against future new values in the enum

File: src/instrumenter/instrument.ts (hunk 1 above)
Comment (cd1m0) [reply src/instrumenter/instrument.ts]: Fixed in [a482f6b](https://github.com/ConsenSys/scribble/pull/154/commits/a482f6b39338dedcd43bdfe3f23939aed396ad48).

File: src/instrumenter/transpiling_context.ts (hunk 2 above)
Comment (cd1m0) [reply src/instrumenter/transpiling_context.ts]: Fixed in [a482f6b](https://github.com/ConsenSys/scribble/pull/154/commits/a482f6b39338dedcd43bdfe3f23939aed396ad48).

File: src/instrumenter/transpiling_context.ts (hunk 3 above)
Comment (cd1m0) [reply src/instrumenter/transpiling_context.ts:225]: Fixed in [a482f6b](https://github.com/ConsenSys/scribble/pull/154/commits/a482f6b39338dedcd43bdfe3f23939aed396ad48).

File: src/instrumenter/transpiling_context.ts [hunk 4]
```diff
@@ -35,11 +35,9 @@ import { makeTypeString } from "./type_string";
...
 
 export enum InstrumentationSiteType {
-    FunctionAnnotation,
-    ContractInvariant,
-    UserDefinedFunction,
-    StateVarUpdated,
-    Assert
+    TwoPointWrapper,
+    SinglePointWrapper,
+    Custom
```
Comment (blitz-1306) [inline src/instrumenter/transpiling_context.ts:40]: I'm wondering if it is possible to document these values to state their meaning and purpose. WDYT?

File: src/spec-lang/tc/typecheck.ts [hunk 5]
```diff
@@ -107,10 +108,38 @@ export type SScope =
...
+function addScope(ctx: STypingCtx, scope: SScope): STypingCtx {
+    return { type: ctx.type, target: ctx.target, scopes: [...ctx.scopes, scope], isOld: ctx.isOld };
+}
+
+function oldCtx(ctx: STypingCtx): STypingCtx {
+    return { type: ctx.type, target: ctx.target, scopes: ctx.scopes, isOld: true };
+}
+
+function fromCtx(ctx: STypingCtx, other: Partial<STypingCtx>): STypingCtx {
+    const t: any = {
```
Comment (blitz-1306) [inline src/spec-lang/tc/typecheck.ts:127]: Not a blocker, but do we get any issues if use `STypingCtx` type instead of `any`?

**Review by blitz-1306 — APPROVED** [review]
Looks good. Two small questions.

File: src/spec-lang/tc/typecheck.ts (hunk 5 above)
Comment (cd1m0) [reply src/spec-lang/tc/typecheck.ts:127]: Yeah - it doesn't type for me. If you can figure out some magic to make it work I'm down?

File: src/instrumenter/transpiling_context.ts (hunk 4 above)
Comment (cd1m0) [reply src/instrumenter/transpiling_context.ts:54]: Fixed in [93c526e](https://github.com/ConsenSys/scribble/pull/154/commits/93c526e31480369c29e5ddbdbf8965f5ea7bfd73)

File: src/spec-lang/tc/typecheck.ts (hunk 5 above)
Comment (cd1m0) [reply src/spec-lang/tc/typecheck.ts]: Fixed in [93c526e](https://github.com/ConsenSys/scribble/pull/154/commits/93c526e31480369c29e5ddbdbf8965f5ea7bfd73)

**Review by blitz-1306 — APPROVED** [review]
Looks good.

**Pull request merged**